src/lib/Generalize.hs

```diff
@@ -0,0 +1,231 @@
+emitGeneralizationParameter :: Type n -> Atom n -> GeneralizerM n (AtomName n)
+emitGeneralizationParameter ty val = GeneralizerM do
+  Abs b v <- return $ newName noHint
+  let emission = Abs (RNest REmpty (GeneralizationEmission (b:>ty) val)) v
+  emitDoubleInplaceTHoisted emission >>= \case
+    Nothing -> error $ "expected atom to be hoistable " ++ pprint val
+    Just v' -> return v'
+
+-- === actual generalization traversal ===
+
+-- Given a type (an Atom of type `Type`), abstracts over all data components
+generalizeType :: Type n -> GeneralizerM n (Type n)
+generalizeType ty = traverseTyParams ty \paramRole paramReqTy param -> case paramRole of
+  TypeParam -> generalizeType param
+  DictParam -> generalizeDict paramReqTy param
+  DataParam -> do
+    paramTyGeneralized <- getType param >>= generalizeType
+    Var <$> emitGeneralizationParameter paramTyGeneralized param
+
+-- === role-aware type traversal ===
+
+-- This traverses the type parameters, with knowledge of their roles and
+-- expected types. It's used here for generalization, but it may also be useful
+-- for other operations on types, like newtype stripping.
+
+traverseTyParams
+  :: (EnvReader m, EnvExtender m)
+  => Atom n
+  -> (forall l . DExt n l => ParamRole -> Type l -> Atom l -> m l (Atom l))
+  -> m n (Atom n)
+traverseTyParams ty f = getDistinct >>= \Distinct -> case ty of
+  TypeCon sn def (DataDefParams ordinaryParams dictParams) -> do
+    Abs roleBinders UnitE <- getDataDefRoleBinders def
+    params' <- traverseRoleBinders f roleBinders (ordinaryParams ++ dictParams)
+    let (ordinaryParams', dictParams') = splitAt (length ordinaryParams) params'
+    return $ TypeCon sn def $ DataDefParams ordinaryParams' dictParams'
+  DictTy (DictType sn name params) -> do
+    Abs paramRoles UnitE <- getClassRoleBinders name
+    params' <- traverseRoleBinders f paramRoles params
+    return $ DictTy $ DictType sn name params'
+  TabPi (TabPiType (b:>(IxType iTy d)) resultTy) -> do
+    iTy' <- f TypeParam TyKind iTy
+    dictTy <- ignoreFallibleT1 $ DictTy <$> ixDictType iTy'
+    d'   <- f DictParam dictTy d
+    withFreshBinder (getNameHint b) (toBinding iTy') \b' -> do
+      resultTy' <- applySubst (b@>binderName b') resultTy >>= f TypeParam TyKind
+      return $ TabTy (b':>IxType iTy' d') resultTy'
+  RecordTy  elems -> RecordTy  <$> traverserseFieldRowElemTypes (f TypeParam TyKind) elems
+  VariantTy (Ext elems Nothing) -> do
+    elems' <- traverse (f TypeParam TyKind) elems
+    return $ VariantTy $ Ext elems' Nothing
+  TC tc -> TC <$> case tc of
+    BaseType b -> return $ BaseType b
+    ProdType tys -> ProdType <$> forM tys \t -> f TypeParam TyKind t
+    SumType  tys -> SumType  <$> forM tys \t -> f TypeParam TyKind t
+    Nat -> return Nat
+    Fin n -> Fin <$> f DataParam NatTy n
+    RefType _ _ -> error "not implemented" -- how should we handle the ParamRole for the heap parameter?
```

Comment:
answering the comment: like a regular type? just abstract over it fully

Reply:
Did you mean "like regular data". I think that's right. Maybe we could treat the heap parameter as a newtype wrapper around `Unit` or something.